### DISCUSSION TITLE
Failing Test (Lesson 12 DSC)

### DISCUSSION THREAD

**Discussion by Amtaf** [Cyfrin/foundry-full-course-cu#819] [Q&A] [answered]
I'm testing "testCanDepositCollateralAndGetAccountInfo()" but i'm getting errors as shown on the screenshot what could be the issue

Here's the test function and my complete project [repo](https://github.com/Amtaf/Algorithmic-StableCoin)
![Screenshot from 2023-09-07 12-21-24](https://github.com/Cyfrin/foundry-full-course-f23/assets/9764833/c8391689-6fda-48e8-b994-b07c39a23931)

```
 function testCanDepositCollateralAndGetAccountInfo() public depositedCollateral{
        (uint256 totalDscMinted, uint256 collateralValueInUsd) = dsce.getAccountInformation(USER);

        uint256 expectedTotalDscMinted = 0;
        uint256 expectedDepositAmount = dsce.getTokenAmountFromUsd(weth, collateralValueInUsd);
        assertEq(totalDscMinted, expectedTotalDscMinted);
        assertEq(AMOUNT_COLLATERAL, expectedDepositAmount);
    }
```

**Comment by Niferu** [comment, marked as answer]
Hello @Amtaf 

In mathematics there are few forbidden operations, one of those is dividing something by 0, it is just not allowed and that is why solidity also does not allow such operations. You have to refactor your all functions that may lead to such math operation to always avoid dividing by 0. Simple make some kind of rule if(x == 0) then revert

**Reply by Amtaf** [reply]
Thank you lemme refactor my functions

**Reply by Niferu** [reply]
@Amtaf 

No problem. If that answers your question please mark it as anser, so otehr with similar question can find it easier, thanks ;)

**Reply by Amtaf** [reply]
Btw thanks ,I checked i lacked a function to calculate the healthFactor that was to Check if the totalDscMinted is equal to 0. If it is, return the maximum value of uint256.

**Reply by Niferu** [reply]
Yup thats the proper solution ;)

**Reply by DukeLurkin** [reply]
What function was that?  I'm still getting the same problem even after copying the repo.

**Reply by DukeLurkin** [reply]
I hope it wasn't :

function _calculateHealthFactor(uint256 totalDscMinted, uint256 collateralValueInUsd)
internal
pure
returns (uint256)
{
if (totalDscMinted == 0) return type(uint256).max;
uint256 collateralAdjustedForThreshold = (collateralValueInUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;
return (collateralAdjustedForThreshold * PRECISION) / totalDscMinted;
}

bc that still is not solving my issue.

**Reply by Niferu** [reply]
@DukeLurkin 

Indeed it is _calculateHealthFactor. Look at mine implementation :)

```solidity
/** @dev It returns max NFR minted value if user total mints is 0, that's why we need to separate it from _healthFactor */
    function _calculateHealthFactor(uint256 totalNfrMinted, uint256 collateralValueInUsd) internal pure returns (uint256) {
        uint256 collateralAdjustedForThreshold = (collateralValueInUsd * LIQUIDATION_THRESHOLD) / LIQUIDATION_PRECISION;

        return totalNfrMinted == 0 ? type(uint256).max : ((collateralAdjustedForThreshold * PRECISION) / totalNfrMinted);
    }
```

Also mine _redeemCollateral has below:
```solidity
if (s_collateralDeposited[from][collateralTokenAddress] == 0 || s_collateralDeposited[from][collateralTokenAddress] < collateralAmount)
            revert NFREngine__NotEnoughCollateralToRedeem();
```
And i got multiple functions with below modifier
```solidity
    modifier moreThanZero(uint256 amount) {
        if (amount == 0) revert NFREngine__NeedsMoreThanZero();
        _;
    }
```

**Reply by DukeLurkin** [reply]
Where in the video did we do any of that?  I'm very confused how I can be missing so many lines of code necessary for this one test to pass.

**Reply by Niferu** [reply]
We did not, it is my solution for a problem

**Reply by DukeLurkin** [reply]
still the same problem after doing all of that.

**Reply by Niferu** [reply]
Hi @DukeLurkin 

You can simple checkout my whole repo: [HERE](https://github.com/Nlferu/NFR_DeFi_Stablecoin)

It has almost 100% test coverage and no errors

**Reply by DukeLurkin** [reply]
@Nlferu thanks so much!  I'll let you know when I get it.
